stdlib/public/core/Collection.swift

```diff
@@ -818,7 +823,7 @@ public protocol Collection: Sequence where SubSequence: Collection {
   ///   a random element.
   /// - Returns: A random element from the collection. If the collection is
   ///   empty, the method returns `nil`.
-  func randomElement<T: RandomNumberGenerator>(
+  __consuming func randomElement<T: RandomNumberGenerator>(
```

Comment:
This surprises me, semantically. `randomElement` seems like it would become rarely useful for move only types. How often does one have a collection and the *only* thing done with it is retrieving a single random element?

This also applies (although slightly less so) to the `SubSequence`-returning methods above: my impression is they're often meant to be views into the larger collection, and having them consume almost defeats the purpose.

It does seem unfortunate to force all of these sorts of accesses through coroutines to be able to "return" borrows, but... I question the value of having these enabled-for-move-only in this way.

Reply:
cc @airspeedswift. Possibly many of these are "bogus customization points" he wants to prune from the protocols. Since we have no concrete plans to introduce "conditional requirements" that only exist for some subset of conforming types, a protocol requirement interface has to be implementable by a future move-only type if it's to remain as a customization point.

Reply:
`first` and `last` are probably bogus. But `randomElement` has a purpose being a protocol requirement... it's the only way we can make `(Int.min...Int.max).randomElement()` work because of the limitation of index distances being an `Int`.

Reply:
OK. Nonetheless, my goal here is not to judge the usefulness of existing protocol requirements as API for working with move-only types, but to make sure the ABI we ship for these protocols with Swift 5 is forward-compatible with implementations we know we want to make move-only. It's inevitable that move-only types will also need new APIs to work with them effectively.

Reply:
In this case, maybe `randomIndex<T: RandomNumberGenerator>(using: inout T) -> Index?` would be a better protocol requirement (since subscripting can be applied to move-only collections), and `randomElement` could be a protocol extension method?

Reply:
I was thinking that these remain non-`__consuming` but implemented in a way that works for move-only types, i.e. "returning" a borrow via a coroutine or whatever our thoughts are on how that works.

Reply:
I don't think we're planning to implement coroutines beyond for property/subscript accessors for Swift 5, though, which limits our short-term options. @rjmccall had thoughts about specializing the borrowed representations of certain types, but there are significant technical challenges to do so, and I'm not sure that such a mechanism would work when abstracted to work generically on an arbitrary collection's `SubSequence`. Without first-class borrowed references in the type system, it seems to me like the primary way you'd have to deal with borrowing subsections of collections is via indexes.

Reply:
Yeah, the idea of consuming an entire collection of move-only types just to extract a single element seems like not something users will generally want.

Reply:
I think conditional requirements are probably going to be a necessity in the future if we can't change the requirement here to be `randomIndex`.

Reply:
Yeah, I started a discussion to that effect on the forums: https://forums.swift.org/t/proposed-amendment-to-se-202-replace-collection-randomelement-requirement-with-randomindex/13056 We might also conceivably drop this requirement altogether; randomElement is far from alone in being the only collection operation that would break on index distance overflow.